src/lib/completion.ts

```diff
@@ -1,131 +1,144 @@
 	}
 
-	/*
-        We need to pass with_vector to qdrant to get our response
-    */
-	private async qdrant_similarity_search(query: string, k: number): Promise<Document[]> {
-		const query_embedding = await this.embeddings_model.embedQuery(query);
-		const qdrant_results = await this.qdrant_client.search(this.qdrant_collection, {
-			vector: query_embedding,
-			limit: k,
-			with_vector: true
+	public async setup() {
+		const tools = getTools();
+		const model = new ChatOpenAI({
+			openAIApiKey: this.openai_api_key,
+			modelName: this.model_name,
+			temperature: 0,
+			stop: ['\nObservation']
 		});
-
-		console.log(qdrant_results);
-
-		// eslint-disable-next-line @typescript-eslint/no-explicit-any
-		return qdrant_results.map((result: any) => {
-			return new Document({
-				pageContent: result.payload.pageContent,
-				metadata: result.payload.metadata
-			});
-		});
-	}
-
-	private async get_vector_response(query: string): Promise<string[]> {
-		console.log('Retrieving vector response from qdrant...');
-
-		const vector_response = await this.qdrant_similarity_search(query, 2);
-
-		console.log(vector_response);
-
-		console.log('Vector response retreived');
-
-		// eslint-disable-next-line @typescript-eslint/no-explicit-any
-		return vector_response.map((document: Document<Record<string, any>>) => {
-			return document.pageContent;
+		const runnable = RunnableSequence.from([
+			{
+				input: (values: InputValues) => {
+					return values.input;
+				},
+				intermediate_steps: (values: InputValues) => {
+					return values.steps;
+				}
+			},
+			this.formatMessages,
+			model,
+			this.customOutputParser.bind(this)
+		]);
+		const executor = new AgentExecutor({
+			agent: runnable,
+			tools: tools
 		});
+		this.executor = executor;
 	}
-
-	private generate_history(history: ChatHistory[]): BaseMessage[] {
-		return history.map((message: { content: string; type: ChatHistoryType }) => {
-			if (message.type == ChatHistoryType.AI) {
-				return new AIMessage({ content: message.content });
-			} else {
-				return new HumanMessage({ content: message.content });
+	private formatMessages = async (values: InputValues) => {
+		const tools = getTools();
+		const intermediateSteps = values.intermediate_steps
+			? (values.intermediate_steps as Array<AgentStep>)
+			: [];
+		const agentScratchpad = formatLogToString(intermediateSteps);
+		let toolString = '';
+		let toolNames = '';
+		for (let i = 0; i < tools.length; i++) {
+			toolString += '\n' + tools[i].name + ': ' + tools[i].description;
+			toolNames += tools[i].name + ', ';
+		}
+		toolNames = toolNames.slice(0, -1);
+		toolNames = toolNames.slice(0, -1);
+		toolString += '\n\n';
+
+		const formatted = [
+			PREFIX,
+			toolString,
+			TOOL_INSTRUCTIONS_TEMPLATE.replace('{tool_names}', toolNames),
+			SUFFIX.replace('{input}', values.input),
+			agentScratchpad
+		].join('');
+		if (this.verbose) {
+			console.log(colors.fg.green, 'Model prompt: ', colors.style.reset);
```

Comment:
nit: We should probably use a proper logging library but if you don't have the bandwidth for this it isn't needing for release so you could ignore this.

Reply:
What do you mean by a proper logging library?

Reply:
Some library that supports things like logging to a file, thats my main thing. I want the ability to have file logs to look at later.

Reply:
Oh, ok I will do it after I finish fine tuning because it is not very urgent right now

Reply:
Actually, I will do logging right now because it is useful to generate data for finetuning.